### ISSUE TITLE
Precompiled LLVM packages for OS X x86-64

### ISSUE THREAD

**Issue by durswd** [mono/CppSharp#694]
##### Brief Description

Compile failed on OS X x86-64.
##### Used settings

```
git clone https://github.com/mono/CppSharp.git
cd CppSharp/build/
./premake5-osx --file=scripts/LLVM.lua download_llvm
./premake5-osx gmake
config=release_x64 make -C gmake
```
##### Stack trace or incompilable generated code

```
Linking CppSharp.CppParser
ld: warning: directory not found for option '-L/usr/lib64'
ld: warning: ld: warning: ld: warning: ld: warning: ignoring file ../../scripts/llvm-cccdd2-macosx-Release/build/lib/libclangFrontend.a, file was built for archive which is not the architecture being linked (x86_64): ../../scripts/llvm-cccdd2-macosx-Release/build/lib/libclangFrontend.aignoring file ../../scripts/llvm-cccdd2-macosx-Release/build/lib/libclangSerialization.a, file was built for archive which is not the architecture being linked (x86_64): ...

ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I think that current llvm library does not contain x86-64 files.
If this is my mistakes, sorry for taking the time and effort.

**Comment by tritao** [comment]
We do not yet provide pre-compiled LLVM packages for x86_64 since Mono on OSX has historically only supported 32-bit. I think the current Mono releases now also support 64-bit with a command line flag, I will have to check.

Maybe we should provide both, but I am not sure if it is possible to support "universal" CppParser parser library that supports both 32/64-bit with same binary.

For now you can follow the guide to compile LLVM manually in 64-bit mode.

I will leave this issue open so we can track this problem.

**Comment by durswd** [comment]
Thank you for reply.
The latest mono for OSX is 64 bit.

I try to compile LLVM manually.

**Comment by Klipi** [comment]
I got stuck with the same issue just now. After compiling LLVM, what's the way to have the build script use the LLVM build under CppSharp/deps instead of the one in CppSharp/build/scripts?

I tried `./premake5-osx --file=scripts/LLVM.lua package_llvm` after quickly skimming through the build script, but that failed with the following output:
`
copy "/Users/(removed)/Documents/CppSharp/CppSharp/deps/llvm/include**" to "llvm-cccdd2-macosx-Release/include".
957 files copied.
copy "/Users/(removed)Documents/CppSharp/CppSharp/deps/llvm/llvm-cccdd2-macosx-Release/include**" to "llvm-cccdd2-macosx-Release/build/include".
Error: .../Documents/CppSharp/CppSharp/build/scripts/Utils.lua:128: /Users/(removed)/Documents/CppSharp/CppSharp/deps/llvm/llvm-cccdd2-macosx-Release/include/ is not an existing directory!
`

I'm not sure if there's something wrong with my build of LLVM, or if I'm using the build script incorrectly.

**Comment by tritao** [comment]
@Klipi The LLVM packaging scripts are expecting some build conventions the same script use.
You can build LLVM using the scripts.

Before starting the process  you will want to remove or set to false `-DLLVM_BUILD_32_BITS=true` in https://github.com/mono/CppSharp/blob/master/build/scripts/LLVM.lua#L264.

Also make sure you have ninja and cmake tools installed.

If the clone steps fails, then update to the latest commit. I just pushed an updated Premake binary which has a shell escaping bug fix that used to make the clone step fail. If you already have LLVM cloned, the step just makes sure its on the right revision.

```
../premake5-osx --file=LLVM.lua clone_llvm
../premake5-osx --file=LLVM.lua build_llvm
../premake5-osx --file=LLVM.lua package_llvm
```

**Comment by Klipi** [comment]
Thanks for the reply! I think the problem was that I built LLVM with the instructions in [Building LLVM](https://github.com/mono/CppSharp/blob/master/docs/BuildingLLVM.md). I guess the clone -> build -> package workflow using premake could be updated there?

I ran into another problem later in the build process and opened up a new issue (#695) for it.

**Comment by tritao** [comment]
This has now been fixed.

You can now specify 64-bits target using: `./premake5-osx --arch=x64 --file=scripts/LLVM.lua download_llvm`.
